### PR TITLE
Do not cache (and do not serve cached) cache manager responses

### PR DESCRIPTION
**Pull request by rousskov**
The fixed bug affected cache manager transactions that were using
/squid-internal-mgr URL path prefix with http(s) URL scheme. It did not
affect transactions that were using legacy cache_object URL scheme.

Stale cache manager responses had their Age response header set to the
number of seconds since Unix epoch. If disk and memory caches were
disabled, then cache manager requests just triggered "found KEY_PRIVATE"
WARNINGs in cache.log (for reasons that remain unclear).

Probably broken since 2011 commit e37bd29 that did not expand
HttpRequest::maybeCacheable() (called cacheable() back then)
PROTO_CACHE_OBJECT check to include /squid-internal-mgr requests.

Also added missing Access-Control-* response headers to cache manager
responses in SMP mode and reduced code duplication related to sending
those headers (which led to them missing in SMP Squids).

### CODE REVIEW

**Review by yadij — DISMISSED** [review]
The bug cited as justification for this would be far better resolved by simply making the cache manager generate the response with appropriate Cache-Control:no-store header. That will allow us in future to fix the cases where it should be some other vale - some only need a must-revalidate, some need private, and some should in future be resumable from cache via ETag.

The secondary issue of receiving traffic from outdated Squid with the above bug should only need a much more limited hack on the server-facing side to make manager responses without Cache-Control implicitly have "no-store" set. Or maybe even just document a suggestion to admin for store_miss ACL rules to workaround it.

File: src/mgr/Inquirer.cc
Comment (rousskov) [inline src/mgr/Inquirer.cc:92]: Note that these Inquerer.cc changes added missing Access-Control-* response headers to cache manager responses in SMP mode.

**Comment by rousskov** [comment]
> @[yadij](https://github.com/yadij) yadij dismissed their [stale review](https://github.com/squid-cache/squid/pull/1185#pullrequestreview-1187630611) 3 days ago

Just to avoid misunderstanding, please note that this PR is still blocked on your review -- dismissing the old review does not (and should not) automatically dismiss the outstanding review _request_. You can do the latter manually in the Reviewers box on the PR page (and adjust the PR labels accordingly).

File: src/CacheManager.h
Comment (yadij) [reply src/CacheManager.h:57]: > IMO, the request-time changes in the PR are valuable on their own and should stay because
> 
>     * The sooner Squid knows that the response is going to be uncachable, the better.

mgr responses being non-cacheable is not guaranteed. As I mentioned earlier there are multiple cases where the response SHOULD be cached. All these request-time changes make it much harder to correct those responses caching in future.

> This knowledge helps reduce cache manager request impact on Store state, which may be very valuable in triage and tests. 

I disagree on the utility. It is certain to add one more case to the existing difficulty troubleshooting why apparently cacheable mgr responses from upstream servers are not being stored. Whereas it is easy to tell from upstream headers when a response is incorrectly made cacheable by them.

> Most clients sending cache manager requests do not tell Squid that the responses are not cachable,

Irrelevant. Cacheability in HTTP is determined by the server. At most a client can request a cache be bypassed and/or refreshed.

>     * Those request-time changes fix a known problem with flags.cachable getting overwritten by maybeCacheable(), reduce code duplication of tricky code, improve debugging, improve code documentation, and make Optional trivially copyable. They are a step forward on many levels, and they are not complex.
> 

Maybe, BUT, none of that extra change is actually fixing the cache manager response bug. It would make review and change documentation a lot better if each of those issue fixes were isolated in their own PRs, or at least PRs scoped more tightly to each particular problematic object.

> > Or maybe even just document a suggestion to admin for store_miss ACL rules to workaround it.
> 
> I agree that warning admins that use affected cache manager requests and sharing this workaround is a good idea.
> Once this PR lands, I can volunteer to post a message to squid-users about it if you think that is the best way to
> document that suggestion.

A post to squid-announce is probably better. That would go to squid-users plus a bunch of others.

For backport I would like the change of generated headers, but the rest is unclear at present.

File: src/RequestFlags.h
Comment (yadij) [reply src/RequestFlags.h:131]: FTR; RequestFlags::noCache should be the value of Cache-Control:no-cache received in request headers. Please remember that in any renaming considerations.

File: src/client_side_reply.cc
Comment (yadij) [inline src/client_side_reply.cc:1531]: IMO this can and should be fixed by this PR. Reversing the if-else like below should be correct.
```
@@ -1528,6 +1528,9 @@ clientReplyContext::identifyStoreObject()

    if (r->flags.noCache) {
        // client sent CC:no-cache which prevents delivering a public/cached object.
        identifyFoundObject(nullptr, "no-cache");
    } else {
        const auto e = storeGetPublicByRequest(r);
        identifyFoundObject(e, storeLookupString(bool(e)));
    }
```

File: src/client_side_request.cc
Comment (yadij) [reply src/client_side_request.cc:1105]: This seems to be a bug fix all of its own, slightly tangled up with the issue of what correct default value for "flags.cacheable" should actually be. I suggest that this part of that pile of issues also should not be fixed here.

File: src/RequestFlags.cc
Comment (yadij) [inline src/RequestFlags.cc:34]: Setting these flags to these values is always valid and previous state does not matter.
Only code setting the inverse values away from these needs the attention a comment brings.
```suggestion
    missCachingDecision = false;
    noCache = true;
```

File: src/internal.cc
Comment (yadij) [reply src/internal.cc:46]: Please return. These lines were ordered by frequency of request type to optimize performance despite the string comparisons. mgr URLs are quite rare in reality compared to the others.

File: src/internal.h
Comment (yadij) [inline src/internal.h:34]: This might be better placed in the class CacheManager API.  I do not insist on this.

File: src/peer_digest.cc
Comment (yadij) [inline src/peer_digest.cc:335]: Setting this flag true is always a risky operation worth documenting.
```suggestion
    // XXX flag currently defaults to false. It should be true to allow collapsing and re-sharing of digests while they are fresh.
    req->flags.missCachingDecision = true;
```

File: src/store_digest.cc
Comment (yadij) [inline src/store_digest.cc:424]: Setting this flag to true is a risky operation worth documenting why it needs to be true
```suggestion
    // XXX flag currently defaults to false. It should be true for ....
    flags.missCachingDecision = true; 
```

**Review by yadij — CHANGES_REQUESTED** [review]

File: src/base/Optional.h
Comment (rousskov) [reply src/base/Optional.h:129]: Done in #1188.

File: src/RequestFlags.h
Comment (rousskov) [reply src/RequestFlags.h:131]: > IMO this type change should be a separate PR

Done in #1189.

File: src/internal.cc
Comment (rousskov) [reply src/internal.cc]: Done in de9d0cf to avoid arguing.

File: src/CacheManager.h
Comment (rousskov) [reply src/CacheManager.h:57]: > Alex: Once this PR lands, I can volunteer to post a message to squid-users about it if you think that is the best way to document that suggestion.

> Amos: A post to squid-announce is probably better. That would go to squid-users plus a bunch of others.

Sounds good, will do.

File: src/internal.h
Comment (rousskov) [reply src/internal.h:34]: Yeah, it is difficult to say for sure. I think this is the best location _for now_ because the function is used, in part, by the code that does not need to know any of the cache manager APIs. This is also where many other similar functions are already located. This function _implementation_ also benefits from being close to its sister functions -- it is easier to keep that code in sync that way.

We can and perhaps should refactor this code to allow individual test functions to be separated into their own modules, but it is best done in a dedicated PR, affecting all of these functions. I would not be surprised if proper refactoring will completely remove this pair of files, but all that is out of scope here.

**Review by rousskov — COMMENTED** [review]
I believe I have addressed all the change requests. The branch code now relies on the recently merged PRs extracted from this PR earlier. No changes to the primary fixes. @yadij, please re-review.

**Comment by rousskov** [comment]
@yadij, just a gentle reminder: This `review-1` PR has been waiting for your review for two weeks.

**Review by yadij — COMMENTED** [review]
Just the ongoing discussion about additional code to add/change. The rest of it looks okay.

File: src/client_side_reply.cc
Comment (rousskov) [reply src/client_side_reply.cc:1531]: > the titular scope/purpose of this PR is to make Squid obey CC:no-cache on manager requests.

I see no such scope/purpose or even a hint at such scope/purpose! This PR is about making new cache manager responses uncachable and preventing previously cached responses from being served from the cache. That is all. This PR changes do not (and should not) care about Cache-Control headers in cache manager requests! What can I change in PR description to clarify that?

Comment (rousskov) [reply src/client_side_reply.cc:1531]: @yadij, please move this bug-fixing PR forward. The current basic disagreement related to this added `XXX` comment should not take weeks to resolve, but it has been more than two weeks since your last response. Again, if it helps, I will remove that `XXX` comment.

Comment (rousskov) [reply src/client_side_reply.cc:1531]: @yadij, please move this bug-fixing `review-1` PR forward. It has been waiting for your review for a month already.

Comment (rousskov) [reply src/client_side_reply.cc:1531]: @yadij, please move this bug-fixing `review-1` PR forward. It has been waiting for your review for two months already.

Comment (yadij) [reply src/client_side_reply.cc:1531]: Issue worked around by [c380732](https://github.com/squid-cache/squid/pull/1185/commits/c38073237e29e0bafb3520c418da57625ad9c5ac).

**Review by yadij — APPROVED** [review]

**Pull request closed without merging**
